[dtensor][2/n] use op overload instead of function schema

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109145
* #109144
* #108264
* #108263
* #108262
* #108261
* __->__ #107306

function schema doesn't provide us anything as we can also get the schema from `op._schema`, include the op directly in op_schema makes easier for sharding prop to do fake execution, and in principle it should also make the hash comparison faster as we don't need to hash the function schema, instead we just hash the `id(op)` which is constant

This PR is just a refactor to include op to OpSchema instead of func schema, no other logic changes

wanchaol committed:
[dtensor][2/n] use op overload instead of function schema

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][2/n] use op overload instead of function schema"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][2/n] use op overload instead of function schema"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][2/n] use op overload instead of function schema"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][2/n] use op overload instead of function schema"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][2/n] use op overload instead of function schema"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][2/n] use op overload instead of function schema"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -116,48 +130,21 @@ class OpSchema:
-    is_inplace: bool = False
-    is_out_variant: bool = False
-
-    def __post_init__(self) -> None:
-        # simple analysis of function schema to determine
-        # if this is an inplace/out variant, it might not
-        # be entirely correct, but it's good enough for now.
-        self.is_inplace = self.func_schema.name[-1] == "_"
-        self.is_out_variant = "out" in self.func_schema.overload_name
```
Isn't this making the check being called every time in every DTensor Op dispatch? Or we cannot cache the value of these two, so using `__post_init__` does not help that much?

wanchaol replied:
This post init just try to analyze the operator we are dispatching over is a inplace op or out variant, but we don't really need this information unless for some certain ops (i.e. when we are checking if the op is inplace or out in pointwise ops), so everytime we form a OpSchema we analyze this become redundant, that's why I am deleting it and put this logic inside the op rules directly.

The caching would cache the op_schema's hash only, and because we don't `hash` these two fields anyways, the cache behavior does not change.

fduwjj reviewed (approved):
Overall, this looks good to me and unblock. Only one question here is that, why do we decide to remove `is_inplace` and `is_out_variant` out of opSchema?

wanchaol committed:
Update on "[dtensor][2/n] use op overload instead of function schema"


function schema doesn't provide us anything as we can also get the schema from `op._schema`, include the op directly in op_schema makes easier for sharding prop to do fake execution, and in principle it should also make the hash comparison faster as we don't need to hash the function schema, instead we just hash the `id(op)` which is constant

This PR is just a refactor to include op to OpSchema instead of func schema, no other logic changes

[ghstack-poisoned]
